railties/lib/rails/commands/credentials/credentials_command.rb

```diff
@@ -100,17 +102,60 @@ def missing_credentials_message
+          @content_path ||= determine_path(:content_path,
+                                           default_path: "config/credentials.yml.enc",
+                                           env_path: "config/credentials/#{environment}.yml.enc")
         end
 
         def key_path
-          options[:environment] ? "config/credentials/#{options[:environment]}.key" : "config/master.key"
+          @key_path ||= determine_path(:key_path,
+                                       default_path: "config/master.key",
+                                       env_path: "config/credentials/#{environment}.key")
+        end
+
+        def determine_path(which, default_path:, env_path:)
+          load_environment_config if environment_specified?
+          config_path = Rails.application.config.credentials[which].to_s.gsub(Rails.root.to_s + "/", "")
+
+          if environment_specified?
+            # Rails.configuration initializes credentials paths based on the existence of credentials files. So,
+            # config_path is the same as default_path when credentials don't yet exist for the specified environment
+            # (happens the first time credentials:edit -e $environment is invoked). We don't want the default path
+            # when an environment has been specified though, so only return config_path if it has been changed by the
+            # user. Otherwise, return env_path.
+            return config_path if config_path != default_path
+            env_path
+          else
+            # Rails.configuration initializes credentials paths based on the existence of credentials files.
+            # Furthermore, Rails and its configuration default to the "development" environment when no environment is
+            # specified, unlike the credentials command, which can run without an environment to edit shared files.
+            # config_path is the same as env_path when credentials already exist for the specified environment, which,
+            # according to Rails.configuration, will be development even if no environment is specified for the
+            # credentials command. We don't want the environment-based path when no environment option was given though,
+            # so only return config_path if it has been changed by the user. Otherwise, return default_path.
+            return config_path if config_path != env_path
+            default_path
+          end
         end
 
         def extract_environment_from_path(path)
           available_environments.find { |env| path.include? env } if path.end_with?(".yml.enc")
         end
 
+        def environment
+          # Explicitly keep credentials env in sync with Rails env, which defaults to dev
+          options[:environment] || "development"
+        end
+
+        def environment_specified?
+          options[:environment].present?
+        end
+
+        def load_environment_config
+          path = Rails.root.join("config/environments/#{environment}.rb")
```

Comment:
I believe we need to initialize the application instead of trying to load the environment config. This could be configured in an initializer and this code would not load it, so I think we need to change `require_application!` call with `require_application_and_environment!`

Reply:
Hey @rafaelfranca, thanks for the review. I've been working on implementing your suggestion, but I've hit a few snags, one of which I'm not sure how I should solve.

The problem is that when I run `credentials:edit`, I want to edit the default `config/credentials.yml.enc`, unless `config.content_path` has been set somewhere (maybe in application.rb or, as you suggested, in an initializer). But when I call `require_application_and_environment!`, it loads the `development.rb` config file, and if `content_path` is set there, I will end up editing the development credentials file instead of the default credentials file.

So I'm looking for ways to work around this. 
1. Is there a way to run initializers without loading environment configs? I could try to write something if not - but is that even something we want? Seems like it could get messy.

2. Another option would be to remove [this bit of code](https://github.com/rails/rails/blob/90d0b42bd8206e942597b64163837287caf7119d/railties/lib/rails/application/configuration.rb#L396-L414) and add `key_path` and `content_path` to the environment config file templates. So your development.rb config would have this in it by default: 
    ```ruby
    config.key_path = config/master.key 
    config.content_path = config/credentials.yml.enc
    ``` 
    and developers could go in and change those to `config/credentials/development.*` (or any other path) if they want to.  
Then the credentials commands would default to operating in the development environment, which I think would also be more consistent with other parts of Rails (e.g., rails server, or rails db:migrate).  
(This would also have the added bonus of letting me simplify some of the logic in the credentials commands, since we would no longer have a config value that changes based on the existence of the file it might be pointing to.)
The downside to this approach is that it would require some adjustments from developers who are already using environment-specific credentials.

3. Maybe we could add path arguments to the credentials commands, so users could just specify which files they want to use? This would remove the complexity of figuring which files to use, but seems like it'd be less user-friendly.

I am certainly open to other suggestions, and I'll keep chipping away at this when I can.

Reply:
I think option 3 makes the most sense, but I need to give this a bit more thought. 🤔